gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -240,73 +178,234 @@ protected Path getPublisherOutputDir(WorkUnitState workUnitState, int branchId)
+
+      builder.add(new PublishGroup(branchId, writerFileSystem, writerOutputDir, publishFileSystem, publisherOutputDir));
     }
+    return builder.build();
   }
 
-  protected void addWriterOutputToExistingDir(Path writerOutputDir, Path publisherOutputDir,
-      WorkUnitState workUnitState, int branchId, ParallelRunner parallelRunner) throws IOException {
-    boolean preserveFileName = workUnitState.getPropAsBoolean(ForkOperatorUtils.getPropertyNameForBranch(
-        ConfigurationKeys.SOURCE_FILEBASED_PRESERVE_FILE_NAME, this.numBranches, branchId), false);
-
-    // Go through each file in writerOutputDir and move it into publisherOutputDir
-    for (FileStatus status : this.writerFileSystemByBranches.get(branchId).listStatus(writerOutputDir)) {
-
-      // Preserve the file name if configured, use specified name otherwise
-      Path finalOutputPath =
-          preserveFileName
-              ? new Path(publisherOutputDir,
-                  workUnitState.getProp(ForkOperatorUtils.getPropertyNameForBranch(
-                      ConfigurationKeys.DATA_PUBLISHER_FINAL_NAME, this.numBranches, branchId)))
-          : new Path(publisherOutputDir, status.getPath().getName());
-
-      LOG.info(String.format("Moving %s to %s", status.getPath(), finalOutputPath));
-      parallelRunner.movePath(status.getPath(), this.publisherFileSystemByBranches.get(branchId),
-          finalOutputPath, Optional.<String> absent());
-    }
+  private CompositeCommitAction getCompositeCommitAction(Iterable<PublishCommand> publishCommands) {
+    return new CompositeCommitAction(Iterables.transform(publishCommands,
+            new Function<PublishCommand, CommitAction>() {
+              @Override
+              public CommitAction apply(PublishCommand publishCommand) {
+                return publishCommand.getCommitAction();
+              }
+            }));
+  }
+
+  private static void markWorkUnitCommitted(WorkUnitState state) {
+    state.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
   }
 
-  private ParallelRunner getParallelRunner(FileSystem fs) {
-    String uri = fs.getUri().toString();
-    if (!this.parallelRunners.containsKey(uri)) {
-      this.parallelRunners.put(uri, this.closer.register(new ParallelRunner(this.parallelRunnerThreads, fs)));
+  private static class CompositeCommitAction implements Action {
+    private final Iterable<CommitAction> commitActions;
+
+    public CompositeCommitAction(Iterable<CommitAction> commitActions) {
+      this.commitActions = commitActions;
+    }
+
+    @Override
+    public void apply() throws Exception {
+      for (CommitAction commitAction : commitActions) {
+        commitAction.apply();
+      }
     }
-    return this.parallelRunners.get(uri);
   }
 
-  @Override
-  public void publishMetadata(Collection<? extends WorkUnitState> states) throws IOException {
-    // Nothing to do
+  private static class CommitAction implements Action {
```

Comment:
Can you rename this class to something like `MarkWorkUnitCommittedAction`, and also add some javadocs for this class.